### PR TITLE
Add Initial Solana implementation

### PR DESCRIPTION
**Pull request by CriesofCarrots**
One note about our implementation: solana requires that stake accounts be created with more than the minimum balance for rent exemption, so that they can't be rent-collected out of existence. We have used this value as the basis of Validator.Details.MinimumAmount. However, only lamports above the minimum rent balance will be activated as stake. Not sure if there is a better way to encode/communicate this.

### CODE REVIEW

**Comment by EnoRage** [comment]
Great start! Can I ask you to: 
- [ ] Add public endpoints to config.yml if they are exists 
- [ ] Run go fmt

File: platform/solana/client.go [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package solana
+
+import (
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+const stakeProgramId  = "Stake11111111111111111111111111111111111111"
```
Comment (EnoRage) [inline platform/solana/client.go]: just run go fmt ./...

Comment (CriesofCarrots) [reply platform/solana/client.go]: Fwiw, running that specific command reformatted coins/coins.go, which broke unit tests

**Comment by CriesofCarrots** [comment]
>     * [x]  Add public endpoints to config.yml if they are exists
> 
>     * [x]  Run go fmt

File: platform/solana/client.go (hunk 1 above)
Comment (EnoRage) [reply platform/solana/client.go]: Also had the same issue sometimes :) 
We will try to solve it

**Comment by EnoRage** [comment]
Can you also add 
```
coin.Solana().Handle:       solana.Init(GetApiVar(coin.SOL)),
```
to the getPlatformMap() function at platform/platform.go

File: platform/solana/client.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+	return r.Account, err
+}
+
+func (c *Client) GetEpochInfo() (epochInfo EpochInfo, err error) {
+	err = c.RpcCall(&epochInfo, "getEpochInfo", []string{})
+	return
+}
+
+func (c *Client) GetMinimumBalanceForRentExemption() (minimumBalance uint64, err error) {
+	err = c.RpcCall(&epochInfo, "getMinimumBalanceForRentExemption", []uint64{4008})
```
Comment (EnoRage) [inline platform/solana/client.go]: ../../platform/solana/client.go:36:19: undefined: epochInfo

Comment (CriesofCarrots) [reply platform/solana/client.go]: 🤦‍♀

**Comment by EnoRage** [comment]
Can you please also ensure that current endpoints are working now:

```
http://localhost:8420/v2/solana/staking/validators
Current response: 

{"code":500,"error":{"error":"invalid character ':' after top-level value: unable to fetch validators list from the registry","meta":{"coin":"solana"}}}

http://localhost:8420/v2/solana/staking/delegations/boot1Z6jb15CLqpaMTn2CxktktwZpRAVAgHZEW6SxQ7

Current response: 

{"code":500,"error":{"error":"Post \"https://api.mainnet-beta.solana.com\": dial tcp 34.82.103.142:443: i/o timeout: Unable to fetch delegations list","meta":{}}}
```
Sorry for the delays, it will be better now

**Comment by CriesofCarrots** [comment]
@EnoRage , it looks to me like these failures are because we haven't updated the assets repo yet. Does that seem right? I created this PR today: https://github.com/trustwallet/assets/pull/1466, and will be populating list.json soon.
Is there any tooling to mock validator data in order to test the blockatlas implementation independently?

**Comment by kolya182** [comment]
> @EnoRage , it looks to me like these failures are because we haven't updated the assets repo yet. Does that seem right? I created this PR today: [trustwallet/assets#1466](https://github.com/trustwallet/assets/pull/1466), and will be populating list.json soon.
> Is there any tooling to mock validator data in order to test the blockatlas implementation independently?

Merged

**Comment by EnoRage** [comment]
> @EnoRage , it looks to me like these failures are because we haven't updated the assets repo yet. Does that seem right? I created this PR today: [trustwallet/assets#1466](https://github.com/trustwallet/assets/pull/1466), and will be populating list.json soon.
> Is there any tooling to mock validator data in order to test the blockatlas implementation independently?

Not yet, currently it is an open issue for us, we will have an option to run and test everything with mocks soon

File: coin/coins.yml [hunk 3]
```diff
@@ -364,6 +364,14 @@
   blockTime: 6000
   sampleAddress: 'HewiDTQv92L2bVtkziZC8ASxrFUxr6ajQ62RXAnwQ8FDVmg'
 
+- id: 501
+  symbol: SOL
+  handle: solana
+  name: Solana
+  decimals: 10
```
Comment (hewigovens) [inline coin/coins.yml]:
```suggestion
  decimals: 9
```

File: platform/solana/client.go [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+const stakeProgramId = "Stake11111111111111111111111111111111111111"
+
+type Client struct {
+	blockatlas.Request
+}
+
+func (c *Client) GetCurrentVoteAccounts() (validators []VoteAccount, err error) {
```
Comment (hewigovens) [inline platform/solana/client.go]: besides staking, can we support `TxAPI interface` too? 
```go
TxAPI interface {
	Platform
	GetTxsByAddress(address string) (TxPage, error)
}
```

Comment (CriesofCarrots) [reply platform/solana/client.go]: Our nodes do not index transactions, so we don't have a TxsByAddress rpc available. Would need a client to poll `getConfirmedBlock` and build a db to support this.

Comment (hewigovens) [reply platform/solana/client.go]: I see, can we return empty list then?

Comment (CriesofCarrots) [reply platform/solana/client.go]: Sure, I'll add that

File: platform/solana/model.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+
+type StakeState struct {
+	State                uint32
+	RentExemptReserve    uint64
+	AuthorizedStaker     [32]byte
+	AuthorizedWithdrawer [32]byte
+	UnixTimestamp        int64
+	LockupEpoch          uint64
+	Custodian            [32]byte
```
Comment (vikmeup) [inline platform/solana/model.go]: do we need all fields to be decoded? can we only included used ones?

File: platform/solana/base.go [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+	"github.com/trustwallet/blockatlas/coin"
+	"github.com/trustwallet/blockatlas/pkg/blockatlas"
+)
+
+type Platform struct {
+	client Client
+}
+
+func Init(api string) *Platform {
+	return &Platform{
```
Comment (EnoRage) [inline platform/solana/base.go]: Can you change to this:
```
p := &Platform{
		client: Client{blockatlas.InitClient(api)},
	}
	p.client.Headers["Content-Type"] = "application/json"
	return p

```

Comment (CriesofCarrots) [reply platform/solana/base.go]: ✔️

**Pull request merged**